core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java

```diff
@@ -404,6 +404,26 @@ public PagedBlockStoreMeta getStoreMetaFull() {
         usedBytesOnDirs.build(), blockOnDirs.build());
   }
 
+  /**
+   * Report all block locations in a pageStoreDirs to master using BlockHeartbeat.
+   */
+  @GuardedBy("getLock().readLock()")
+  public void reportBlocks() {
+    final List<PageStoreDir> pageStoreDirs = getStoreDirs();
+    for (PageStoreDir pageStoreDir : pageStoreDirs) {
+      final PagedBlockStoreDir pagedBlockStoreDir = downcast(pageStoreDir);
+      Set<PagedBlockMeta> blockMetas = mBlocks.getByField(INDEX_STORE_DIR, pagedBlockStoreDir);
+      for (PagedBlockMeta blockMeta : blockMetas) {
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          synchronized (listener) {
+            listener.onMoveBlockByWorker(blockMeta.getBlockId(),
+                blockMeta.getBlockLocation(), blockMeta.getBlockLocation());
```

Comment:
Presumably there are far more blocks than event listeners. It's more efficient to only acquire the lock once per listener.
```suggestion
      for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
        synchronized (listener) {
          for (PagedBlockMeta blockMeta : blockMetas) {
            listener.onMoveBlockByWorker(blockMeta.getBlockId(),
                blockMeta.getBlockLocation(), blockMeta.getBlockLocation());
```